solidity/contracts/isms/hook/OPL2ToL1CcipReadIsm.sol

```diff
@@ -0,0 +1,178 @@
+                _message
+            );
+
+        revert OffchainLookup(
+            address(this),
+            urls,
+            ccipReadCallData,
+            OPL2ToL1CcipReadIsm.process.selector,
+            _message
+        );
+    }
+
+    /// @dev called by the relayer when the off-chain data is ready
+    function process(
+        bytes calldata _metadata,
+        bytes calldata _message
+    ) external {
+        mailbox.process(_metadata, _message);
+    }
+
+    function verify(
+        bytes calldata _metadata,
+        bytes calldata _message
+    ) external override returns (bool) {
+        if (_areWeMessageRecipient(_message)) {
+            _proveWithdrawal(_metadata, _message);
+        } else {
+            _finalizeWithdrawal(_metadata, _message);
+        }
+
+        return true;
+    }
+
+    function handle(
+        uint32 _origin,
+        bytes32 _sender,
+        bytes calldata _messageBody
+    ) external payable {
+        emit ReceivedMessage(_origin, _sender, msg.value, _messageBody);
+    }
+
+    function interchainSecurityModule()
+        external
+        view
+        override
+        returns (IInterchainSecurityModule)
+    {
+        return IInterchainSecurityModule(address(this));
+    }
+
+    function _areWeMessageRecipient(
+        bytes calldata _message
+    ) internal view returns (bool) {
+        return _message.recipientAddress() == address(this);
+    }
+
+    function _proveWithdrawal(
+        bytes calldata _metadata,
+        bytes calldata /* _message */
+    ) internal {
+        (
+            IOptimismPortal.WithdrawalTransaction memory _tx,
+            uint256 _disputeGameIndex,
+            IOptimismPortal.OutputRootProof memory _outputRootProof,
+            bytes[] memory _withdrawalProof
+        ) = abi.decode(
+                _metadata,
+                (
+                    IOptimismPortal.WithdrawalTransaction,
+                    uint256,
+                    IOptimismPortal.OutputRootProof,
+                    bytes[]
+                )
+            );
+
+        bytes32 withdrawalHash = OPL2ToL1Withdrawal.hashWithdrawal(_tx);
+
+        // Proving only if the withdrawal wasn't
+        // proven already by this contract
+        if (!_isWithdrawalProvenAlready(withdrawalHash)) {
+            opPortal.proveWithdrawalTransaction(
+                _tx,
+                _disputeGameIndex,
+                _outputRootProof,
+                _withdrawalProof
+            );
+        }
+    }
+
+    /// @dev we handle the access to the provenWithdrawal mapping
+    /// for OptimismPortal version 1 and version 2
+    function _isWithdrawalProvenAlready(
+        bytes32 _withdrawalHash
+    ) internal view returns (bool) {
+        try opPortal.provenWithdrawals(_withdrawalHash) returns (
+            IOptimismPortal.ProvenWithdrawal memory provenWithdrawal
+        ) {
+            return provenWithdrawal.timestamp > 0;
+        } catch {
```

Comment:
i'm not a fan of this try catch for handling multiple versions
can we have an immutable version in the constructor we switch on here?
usually I would suggest having two different implementations but don't want to delay any further